src/puppetlabs/puppetdb/queue.clj

```diff
@@ -30,15 +32,78 @@
     (catch Exception e
       (throw+ {:kind ::parse-error} e "Error parsing command"))))
 
+(defn sanitize-certname
+  "Replace any underscores and filename forbidden characters found in `certname`
+  with dashes."
+  [certname]
+  (let [forbidden-chars (conj constants/filename-forbidden-characters
+                              "_")]
+    (str/replace
+      certname
+      (re-pattern (str "("
+                       (->> forbidden-chars
+                           (map #(format "\\Q%s\\E" %)) ; escape regex chars
+                           (str/join "|"))
+                       ")"))
+      "-")))
+
+(defn max-certname-length
```

Comment:
I highly doubt that operating on small strings that can fit entirely in the L1 cache is ever going to be a CPU bottleneck when there's disk & network IO going on, nor do I see how they're going to exert significant memory pressure when the average command body is way larger. I'd prefer to keep this as clear as possible until I see some data or a compelling argument that is ever going to be more than a blip on CPU or memory usage.